File: leetcode/next-greater-element-i.py
from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        d = {}

        for n in nums2:
            while stack and stack[-1] < n:
                d[stack[-1]] = n
                stack.pop()

            stack.append(n)

        return [d.get(n, -1) for n in nums1]

File: leetcode/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        map<int, int> greater;
        stack<int> st; // monotonic
        vector<int> res;

        res.reserve(nums1.size());

        // construct greater map by monotonic stack
        for(auto n : nums2) {
            while(!st.empty() && st.top() < n) {
                greater[st.top()] = n;
                st.pop();
            }

            st.push(n);
        }

        for(auto n : nums1) {
            if (greater.find(n) == greater.end()) {
                res.push_back(-1);
            } else {
                res.push_back(greater[n]);
            }
        }


        return res;
    }
};